paderbox/array/kernel.py

```diff
@@ -0,0 +1,212 @@
+import typing
+import numpy as np
+import paderbox as pb
+
+
+__all__ = [
+    'np_kernel',
+    'ai_dilate',
+    'ai_erode',
+    'max_kernel',
+    'min_kernel',
+    'mean_kernel',
+    'median_kernel',
+]
+
+
+def _plot(a):
+    """
+    >>> _plot([0, 1, 2, 3, 4, 5, 6, 7, 8])
+    ' ▁▂▃▄▅▆▇█'
+    """
+    a = np.array(a)
+    values = np.array(list(' ▁▂▃▄▅▆▇█'))
+    m = np.amax(a)
+    if m == 0:
+        m = 1
+    a = np.round(a * ((len(values)-1) / m)).astype(int)
+    print(repr(''.join(values[a])))
+
+
+def np_kernel1d(
+        x,
+        kernel_size,  # an odd number
+        *,
+        kernel: callable,  # e.g. np.mean, np.amax, np.amin
+        padding_mode: 'typing.Literal["edge"]' = 'edge',
+        pad_position: 'typing.Literal["pre", "post", None]' = 'pre',
+):
+    """
+    Apply a kernel to the last axis of x.
+
+    Args:
+        x: np.array
+        kernel_size:
+        kernel: a function that should be applied to the kernel window.
+            Has to accept an axis argument, like np.mean, np.amax, np.median.
+        padding_mode:
+            The mode for np.pad, see np.pad for the options.
+        pad_position:
+            'pre': First pad and then apply the kernel.
+            'post': First apply the kernel then pad.
+            None: Don't pad and return an array with a smaller last axis.
+
+    Returns:
+
+    >>> a = np.array([0, 0, 0, 0, 1, 1, 1, 1, 0, 0, 0, 0,])
+    >>> _plot(a); _plot(np_kernel1d(a, 3, kernel=np.amax))
+    '    ████    '
+    '   ██████   '
+    >>> _plot(a); _plot(np_kernel1d(a, 3, kernel=np.amin))
+    '    ████    '
+    '     ██     '
+    >>> _plot(a); _plot(np_kernel1d(a, 3, kernel=np.mean))
+    '    ████    '
+    '   ▃▅██▅▃   '
+
+    """
+    assert kernel_size % 2 == 1, (kernel_size, 'kernel size has to be odd.')
+    assert pad_position in ['pre', 'post', None], pad_position
+
+    if pad_position == 'pre':
+        shift = kernel_size // 2
+        padding = [(0, 0)] * (x.ndim - 1) + [(shift, shift)]
+        x = np.pad(x, padding, padding_mode)
+
+    y = kernel(pb.array.segment_axis(x, kernel_size, 1, end='pad'), axis=-1)
+
+    if pad_position == 'post':
+        shift = kernel_size // 2
+        padding = [(0, 0)] * (y.ndim - 1) + [(shift, shift)]
+        y = np.pad(y, padding, padding_mode)
+    return y
+
+
+def _ai_dilate_erode(ai, khalf):
+    """
+
+
+    # Test inverse mode an unknown shape
+    >>> a = pb.array.interval.ones()
+    >>> a[3:7] = False
+    >>> _plot(a[:10]); _ = [_plot(ai_dilate(a, k)[:10]) for k in [1, 3, 5, 11]]
+    '███    ███'
+    '███    ███'
+    '████  ████'
+    '██████████'
+    '██████████'
+
+    """
+    pairs = np.array(ai.normalized_intervals)
+    assert pairs.shape[-1] == 2, pairs.shape
+    if ai.inverse_mode:
+        khalf = -khalf
+    pairs[:, 0] -= khalf
+    pairs[:, 1] += khalf
+    pairs = np.maximum(pairs, 0)
```

Comment:
Should this be done in `from_pairs`? It currently fails with negative indices and I'm not sure if it makes sense to clip to 0 or if it's better to fail in general.

Reply:
This shouldn't be done in `from_pairs`. Negative indices in python count from the end.
`from_pairs` can be interpreted as an alternative of
```python
ai = zeros(shape=...)
for s,e in pairs:
    ai[s:e] = True
```

We may want to change `from_pairs` to fail for negative indices
```python
>>> ArrayInterval_from_pairs([[-4, 2]], shape=5)
ArrayInterval("1:2", shape=(5,))
```
But that is an independent discussion.